### DISCUSSION TITLE
Enable RichText spell checking

### DISCUSSION THREAD

**Discussion by MaRoe** [EclipseNebula/nebula#572] [Q&A] [answered]
I've been struggling for a couple of hours to get the spell checking plugin of the CKEditor working in an RCP application?

The plugin is available in the editor package - does anyone has an idea if it is possible to enable it?

Cheers and regards
Martin

**Comment by lcaron** [comment]
Hi martin

Can you please try to change the "template" file [1] by replacing the `<html>` tag by `<html lang="fr" spellcheck="true">` ? 

Laurent

[1] = https://github.com/eclipse/nebula/blob/master/widgets/richtext/org.eclipse.nebula.widgets.richtext/src/org/eclipse/nebula/widgets/richtext/resources/template.html

**Comment by MaRoe** [comment]
Hi Laurent,

thanks for the blazing fast response. Unfortunately adding the attribute does not enables the spellchecker.

Cheers
Martin

**Comment by lcaron** [comment]
Hi Martin

Ok sad news. I'll have a look tomorrow (tonight is dedicated to the Dune 2 movie :) )
Can you please tell me on which operating system do you try the RTE component ?

Cheers

Laurent

**Reply by MaRoe** [reply]
Hi Laurent, Hope your Dune session was terrific :)

@fipro78 checked on the the issue (see below) and seems to run into the same issue as me. Do you have any additional ideas?

**Comment by MaRoe** [comment]
Hi Laurent,

yeah, everyone has to set the right priorities - Have fun :)

I am running on 
- Windows Server 2019 Datacenter, Build 1809
- Java Temurin 1.8 382
- RichText Version 1.5.0.20220822171809
- Eclipse Neo.3 Release (4.6.3)

Cheers
Martin

**Reply by fipro78** [reply]
From reading the docs I would assume you need to add the option `disableNativeSpellChecker` with value `false` to the `RichTextEditorConfiguration`. 

But I haven't tried it before, so it is just a guess. 

https://ckeditor.com/docs/ckeditor4/latest/features/spellcheck.html

**Reply by fipro78** [reply]
I tried it this morning, but with no success. All the options described in the documentation of various sites have no effect. Can't tell why it is not working. The plugins should be loaded and available.

**Reply by lcaron** [reply]
Hi dirk

Do you want me to have a look ?

Laurent

**Reply by fipro78** [reply]
@lcaron if you want, feel free. Maybe you have an idea. I tried all that I could think of. For me it does not make sense that the jar is not loaded and activated. But maybe it is related to the swt browser. I recently noticed that Javascript is sometimes not loaded if some incompatible syntax is used. At least I noticed that on my windows 11. But I have no idea about the reason.

**Reply by MaRoe** [reply]
@lcaron Sorry to bother you, but did you have had time to look onto this issue? I really would like to add the "native" editor spellchecking to our product, but our customer is getting somewhat impatient ;)

**Comment by lcaron** [comment]
Hi Martin

Sorry for my late answer, time flies like an arrow. I'll work on this topic this afternoon. I'll keep you informed

**Comment by lcaron** [comment]
Hi martin

I think that I found something : 
* In the file "template.html" [1] change the `<html>` tag by `<html spellcheck="true">`.
* In the file "config.js" [2] add the parameter `config.disableNativeSpellChecker = false;`

```
CKEDITOR.editorConfig = function( config ) {
	// Define changes to default configuration here. For example:
	// config.language = 'fr';
	// config.uiColor = '#AADC6E';
	config.disableNativeSpellChecker = false;
};
```

![image](https://github.com/eclipse/nebula/assets/1170014/5ce63b05-ea33-4642-a995-dd256444e633)

It is OK for you, we can add an enhancement.

Laurent

[1] = https://github.com/eclipse/nebula/blob/master/widgets/richtext/org.eclipse.nebula.widgets.richtext/src/org/eclipse/nebula/widgets/richtext/resources/template.html

[2] = https://github.com/eclipse/nebula/blob/master/widgets/richtext/org.eclipse.nebula.widgets.richtext/src/org/eclipse/nebula/widgets/richtext/resources/ckeditor/config.js

**Reply by fipro78** [reply, marked as answer]
Hi @lcaron,

nice that you found it. Actually we would only need to add `<html spellcheck="true">` to the template.html.

Whether a user wants to support spellchecking or not, can then be configured via the `RichTextEditorConfiguration`:

```
RichTextEditorConfiguration config = new RichTextEditorConfiguration();
config.setOption("disableNativeSpellChecker", Boolean.FALSE);
final RichTextEditor editor = new RichTextEditor(parent, config);
```

There is no need to modify the config.js, which is actually a file provided by ckeditor and I haven't touched for license obligations. I don't know if that would be an issue, but I didn't wanted to touch the provided files. Actually this is one of the reasons for the `RichTextEditorConfiguration`, besides the fact that a user wants to customize the config without the need to change files in the jar.

For convenience we could add a constant and a method to the `RichTextEditorConfiguration` to make it easier to enable the spellchecking. I would then go for "enable" rather than "disable", as from an API point of view it is easier to enable something, instead of not disabling something. ;-)

```
/**
 * Key to configure whether to enable the resizing feature.
 * @since 1.6
 */
public static final String DISABLE_NATIVE_SPELL_CHECKER = "disableNativeSpellChecker";

/**
 * Whether to enable the native spell checking feature.
 *
 * @param enable
 *            <code>true</code> to enable the native spell checking.
 *  @since 1.6
 */
public void setEnableNativeSpellChecker(boolean enable) {
	this.options.put(DISABLE_NATIVE_SPELL_CHECKER, !enable);
}
```

As it would be a new API, the version of the RichText plugin needs to be updated to 1.6 with the convenience API.

The minimum fix without API change would be only the modification of the `html` tag in the template.html. Although in my tests today, even only adding the option to the `RichTextEditorConfiguration` works. Not sure if I am facing a caching issue.

Maybe @MaRoe could first test with only the `config.setOption("disableNativeSpellChecker", Boolean.FALSE);` variant, and then with adding the `spellcheck` attribute to the html tag. Maybe adding the attribute could even work dynamically using Javascript, e.g.

```
config.getBrowser().addProgressListener(new ProgressListener() {
	
	@Override
	public void completed(ProgressEvent event) {
		editor.executeJavascript("document.documentElement.setAttribute('spellcheck', 'true');");
		// editor.executeJavascript("alert(document.documentElement.getAttribute('spellcheck'));");
	}
			
	@Override
	public void changed(ProgressEvent event) {
	}
});
```

Then we know what we need to add at least in the code.

**Reply by MaRoe** [reply]
Hi there and thank you both for your assistance.
After doing it like suggested everything works like charm.

I encountered a strange behaviour that prevents the actual spellchecking - which isn't related to Nebula directly but I guess with the internal SWT Browser and I want to document it because somebody might come across the same issue ;)
My default windows language is German and I added English as default App language as I use a keyboard with American layout. My Nebula editor spellchecker is configured to check German. If my Keyboard layout (and the App language) is set to English the spell checking does not work (which seems logical as an embedded Windows browser is utilized afaik).

**Reply by fipro78** [reply]
Which steps were necessary at least to make it work? Just so that we know what to add to the sources to make it easier to enable spellchecking in future versions.

**Reply by MaRoe** [reply]
1. I changed the html-tag in template.html the following way `<html lang="de" spellcheck="true">`
2. Added `getEditorConfiguration().setLanguage(Locale.GERMAN)` as fallback to point 1
3. Added `getEditorConfiguration().setOption("disableNativeSpellChecker", false)`

This made the spellchecking work.

**Comment by fipro78** [comment]
@lcaron 
So it looks like we at least should add `<html spellcheck="true">` to the template.html. Since you found out that part, do you want to change this?

And what do you think about the convenience API? Should we add it, or just leave it the way it is?

**Reply by lcaron** [reply]
Hi @fipro78 
I'm gonna handle the changes (HTML, API, Increase version number).

Thank you everybody,

Laurent
